content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+| Python       | 60.71       | 36.20    | 134.55    | 51.90    | 432.73   | 591.61   |
+| Java         | 56.94       |  5.17    |  94.11    | 39.02    | 266.93   | 489.04   |
+| Javascript   | 58.21       | 57.11    | 139.24    | 32.77    | 387.57   | 449.16   |
+| Ruby         | 37.78       | 13.97    |  48.80    | 23.36    | 151.13   | 193.24   |
+| Go           | 80.89       | 49.41    | 110.96    | 41.23    | 326.09   | 467.81   |
+| All languages| 60.41       | 37.59    | 116.23    | 37.37    | 336.30   | 459.77   |
+
+## Maneuvering
+
+The rest of the apollo pipeline ran much smoother. Thanks to [MinhashCuda](https://github.com/src-d/minhashcuda)
+the hashing stage took less then a day to finish. Our previous [blog post](../minhashcuda)
+can provide more details about MinhashCuda. It particularly describes the
+Locality Sensitive Hashing procedure in depth - the same that we employ for the deduplication.
+In a few words, we scan the buckets of the hash tables and treat the files
+in the same bucket as similar and those files form a clique.
+Hence the computational complexity is linear.
+See also [ekzhu/datasketch](https://ekzhu.github.io/datasketch/lsh.html).
+
+We decided to use two similarity thresholds for the hashing process, a stricter 95% 
+and looser 80%, the same the *DéjàVu* authors had used in `SourcersCC`. They yielded
+the following results after the pairwise similarity graph was split into connected
+components (CCs):
+
+|                                      |       80% threshold        |         95% threshold        |
+|--------------------------------------|----------------------------|------------------------------|
+|Cliques count                         | 7,482,339                  | 7,732,877                    |
+|CCs count                             | 3,473,933                  | 6,483,934                    |
+|Count for CCs with 1 file             | 3,000,759 (38.13% of files)| 6,211,433 (78.93%  of files) |
+|Count for CCs with >1 file            | 473,174 (61.87% of files)  | 1,496,749 (21.07% of files)  |
+|Avg. files number per CC with >1 file | 10.29                      | 6.09                         |
+|Maximum files number across all CCs   | 223,309                    | 6,839                        |
+
+How do we interpret these results? If there were few exact clones, the number
+of cliques would have been much lower, and the difference in those numbers
+for the two thresholds would have been more than 3%.
+Even though the histograms below show that the majority of the connected components
+have a small number of files (distributed exponentially), the amount of 
+duplication that entails in GitHub's most starred repositories is astonishing.
+
+{{% caption src="/post/deduplicating_pga_with_apollo/hist.png" %}}
+Log-log histograms of the number of distinct file names in CCs, at the 80% threshold (left)
+and the 95% threshold (right).
+{{% /caption %}}
```

Comment:
the image is not clickable to enlarge it and the basic font is too small/too blurry.

Reply:
increased font